torchgen/gen_functionalization_type.py

```diff
@@ -351,6 +334,95 @@ def emit_view_functionalization_body(
+        return ''
+    return_type = dispatcher.returns_type(f.func.returns).remove_const_ref().cpp_type()
+    return f'{return_type} {var_name} = '
+
+def wrap_propagate_mutations_and_return(f: NativeFunction, functional_op: Optional[NativeFunction], inner_out_name: str) -> str:
+    if len(f.func.returns) > 1:
+        # Some assertions: We don't want any functions with a return type of "-> (Tensor(a!), Tensor)",
+        # because:
+        # (1) It's more annoying to handle properly
+        # (2) It's unnecessary - you can't method-chain on the first (mutated) output because it's part of a tuple.
+        # Instead, we expect the (a!) argument to not be returned.
+        mutable_rets = [r for r in f.func.returns if r.annotation is not None and r.annotation.is_write]
+        immutable_rets = [r for r in f.func.returns if r.annotation is None]
+        assert len(mutable_rets) == 0 or len(immutable_rets) == 0, f"""\
+Found a mutating operator ({f.func.name}) which returns multiple outputs, some of which alias the inputs and some that do not.
+ Instead, only the non-aliased outputs should be a part of the return type (and the mutations will happen by side effect)."""
+
+    # And with the above assumption, we can assume that returns are either aliased or not.
+    returns_are_aliased = any(r.annotation is not None for r in f.func.returns)
+    updates = []
+    return_names = []
+
+    if functional_op is None:
+        if returns_are_aliased:
+            # Returns are aliased, so we can just return the original inputs
+            for r in f.func.returns:
+                aliased_name = next(a for a in f.func.arguments.flat_all
+                                    if a.annotation is not None and a.annotation == r.annotation)
+                return_names.append(aliased_name.name)
+        else:
+            # Need to wrap and return the fresh outputs
+            return_is_tuple = len(f.func.returns) > 1
+            for (i, r) in enumerate(f.func.returns):
+                inner_ret: str = f'std::get<{i}>({inner_out_name})' if return_is_tuple else inner_out_name
+                updates.append(f"""\
+      auto output_{i} = at::functionalization::impl::to_functional_tensor({inner_ret});""")
+                return_names.append(f'output_{i}')
+    else:
+        # More assertions:
+        # We have some native functions that mutate their inputs, and also return fresh outputs
+        # (this applies to "running-average" type operators, like batch norm).
+        # If an operator has a schema like this:
+        #     foo(Tensor(a!) input) -> Tensor
+        # Then if a corresponding out-of-place version exists, we expect it to have the following schema:
+        #     foo.functional(Tensor input) -> (Tensor, Tensor)
+        # The first tensor(s) in the output should all correspond to the newly updated inputs.
+        # The last tensor(s) in the output should correspond to the original outputs of the function.
```

Comment:
So... I don't know of any easy way to assert the "first / last" part of this comment. The assert below only checks that the # of returns is correct.

I think I'm ok with this though mostly because you would only add one of these functional ops in the first place because you wanted it to work with functionalization, and if you returned arguments in the wrong order then functionalization would immediately give you garbage results.

I guess if we added first class support for OpInfo <> functionalization testing, and just ran it automatically on all OpInfos, then we could catch it.

Reply:
The fix is easy: put names in the returns.

Reply:
Well technically, that still doesn't prevent the person implementing the kernel from returning a tuple of tensors with some of the tensors swizzled around (not obeying whatever ordering they implied in the schema based on the names).

The names we give to the returned tensors in the schema can't be the same as the input tensors - since they're fresh, non-aliased tensors (although we could maybe enforce some convention like "`foo.functional(Tensor input1, Tensor inpt2) -> (Tensor inpt1_out, Tensor inpt2_out, Tensor)`

Reply:
> Well technically, that still doesn't prevent the person implementing the kernel from returning a tuple of tensors with some of the tensors swizzled around (not obeying whatever ordering they implied in the schema based on the names).

Well, sure, but there's no way to solve this. I'll settle for making sure the functional and regular variants agree haha.

> The names we give to the returned tensors in the schema can't be the same as the input tensors - since they're fresh, non-aliased tensors (although we could maybe enforce some convention like "foo.functional(Tensor input1, Tensor inpt2) -> (Tensor inpt1_out, Tensor inpt2_out, Tensor)

A naming convention sounds like the right thing here.